fire_atlas_editor/lib/screens/editor_screen/widgets/selection_canvas/canvas_board.dart

```diff
@@ -64,37 +70,42 @@ class CanvasBoardState extends State<CanvasBoard> {
   }
 
   void _handleMove(DragUpdateDetails details) {
-    setState(() {
-      final x = details.localPosition.dx - _lastDrag.dx;
-      final y = details.localPosition.dy - _lastDrag.dy;
-
-      if (_currentTool == CanvasTools.MOVE) {
-        _translateX += x;
-        _translateY += y;
-      } else if (_currentTool == CanvasTools.SELECTION) {
-        _finishSelection(_calculateIndexClick(details.localPosition));
-      }
-
-      _lastDrag = details.localPosition;
-    });
+    if (_lastDrag != null) {
+      final _last = _lastDrag!;
+      setState(() {
+        final x = details.localPosition.dx - _last.dx;
+        final y = details.localPosition.dy - _last.dy;
+
+        if (_currentTool == CanvasTools.MOVE) {
+          _translateX += x;
+          _translateY += y;
+        } else if (_currentTool == CanvasTools.SELECTION) {
+          _finishSelection(_calculateIndexClick(details.localPosition));
+        }
+
+        _lastDrag = details.localPosition;
+      });
+    }
   }
 
   void _handleMoveEnd() {
-    setState(() {
-      if (_currentTool == CanvasTools.SELECTION) {
-        _finishSelection(_calculateIndexClick(_lastDrag));
-      }
-      _lastDrag = _dragStart = null;
-    });
+    if (_lastDrag != null) {
+      setState(() {
+        if (_currentTool == CanvasTools.SELECTION) {
+          _finishSelection(_calculateIndexClick(_lastDrag!));
+        }
+        _lastDrag = _dragStart = null;
+      });
+    }
   }
 
   void _handleMoveStart(DragStartDetails details) {
     setState(() {
       _lastDrag = _dragStart = details.localPosition;
 
       if (_currentTool == CanvasTools.SELECTION) {
-        _selectionStart = _calculateIndexClick(_dragStart);
-        _finishSelection(_selectionStart);
+        _selectionStart = _calculateIndexClick(_dragStart!);
+        _finishSelection(_selectionStart!);
```

Comment:
Calculating and calling methods inside a `setState` should be prevented if possible, it can have unexpected side effects.

Calculate it before, and trigger a state update after you are sure everything is done.

Reply:
hmmm, that seems to make sense, I just curious on what side effects this could led to.

Changed anyway

Reply:
multiple state rebuilds because the called method calls setState again, is an example I can think off from the top of my head

Reply:
I see, I don't think that would be case on this code. But this is a good rule to use.